[o2k] copy plugin objects as is (and respect enabled flag)

dimitropoulos opened the pull request:
closes: INS-648
note: this is stacked off of https://github.com/Kong/insomnia/pull/3416

@Tieske please take a look at https://github.com/Kong/insomnia/pull/3419/commits/d21787c2112c931e5458585c7027ae56c08e85d1 which respects the enabled flags and https://github.com/Kong/insomnia/pull/3419/commits/f27bc298f0eca06fe6732b709dba7ec0a8dbd8ee which allows passing more than just `name`, `config`, and `tags`.

If those are correct, the only remaining question is whether it's also correct that we should be adding tags into the resulting plugin config.  Presently we do, and the current test (intentionally, for now, until I hear from you) fails on this:

<img height="300" src="https://user-images.githubusercontent.com/15232461/119394073-c96b3380-bc9f-11eb-82f4-1145643c1201.png" />

dimitropoulos committed:
clears all errors in tests, enables overrides for spec generators

dimitropoulos committed:
consolidates openapi x-kong extension to one place, uses consistently

these were used for some things, not used for others, exported in some places, not exported in others.  now it's uniform and we can look up references for each type.

dimitropoulos committed:
removes copy/pasta in tests, uses spec generator, removes useless async

continuing from the last commit, but separated out for easier review

dimitropoulos committed:
adds failing test for ticket

dimitropoulos committed:
only adds `enabled` where it was specified

dimitropoulos committed:
allows adding passing top-level properties to config

now, the line that was there specifically for `config` is redundant.

Tieske reviewed (commented):
lgtm

- the `enabled` property should not be generated. If it is specified in the input, take that, if not then do not add it. Result is that the user specified value is honored, and if not specified, it will fall back on Kong defaults. So this part is ok.
- the `tags` property should be added to the plugin config (as for each Kong entity). So the code is right, and the test needs updating.

dimitropoulos committed:
adds tags to test result per review feedback

dimitropoulos commented:
ok, I added the tags like you said in 4ccc12909289f9c24a80782d99ee6f53ff21430f

dimitropoulos committed:
removes unnecessary generic

for some reason, this didn't error before, but now it is.

dimitropoulos committed:
removes unnecessary argument

this failed on the CI with:

openapi-2-kong: FAIL src/generate.test.ts
openapi-2-kong:   ● Test suite failed to run
openapi-2-kong:     src/generate.test.ts:90:58 - error TS2769: No overload matches this call.
openapi-2-kong:       Overload 1 of 2, '(src: string, options?: (ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined): any', gave the following error.
openapi-2-kong:         Argument of type '"utf8"' is not assignable to parameter of type '(ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined'.
openapi-2-kong:       Overload 2 of 2, '(src: string, reviver: Reviver, options?: (ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined): any', gave the following error.
openapi-2-kong:         Argument of type 'string' is not assignable to parameter of type 'Reviver'.
openapi-2-kong:     90       const parsedSpec = YAML.parse(dcFixtureFileString, 'utf8');
openapi-2-kong:                                                                 ~~~~~~
openapi-2-kong:     src/generate.test.ts:100:59 - error TS2769: No overload matches this call.
openapi-2-kong:       Overload 1 of 2, '(src: string, options?: (ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined): any', gave the following error.
openapi-2-kong:         Argument of type '"utf8"' is not assignable to parameter of type '(ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined'.
openapi-2-kong:       Overload 2 of 2, '(src: string, reviver: Reviver, options?: (ParseOptions & DocumentOptions & SchemaOptions & ToJSOptions) | undefined): any', gave the following error.
openapi-2-kong:         Argument of type 'string' is not assignable to parameter of type 'Reviver'.
openapi-2-kong:     100       const parsedSpec = YAML.parse(k8sFixtureFileString, 'utf8');

develohpanda commented on packages/openapi-2-kong/src/declarative-config/plugins.ts:
```diff
@@ -147,10 +147,16 @@ export function generateRequestValidatorPlugin({ plugin, tags, operation }: {
     config,
-    tags: [...(tags ?? []), ...(plugin.tags ?? [])],
-    enabled: Boolean(plugin.enabled ?? true),
+    tags: [
+      ...(tags ?? []),
+      ...(plugin.tags ?? []),
+    ],
+    ...enabled,
```
I wonder if we should skip tags entirely if both `tags` and `plugin.tags` are empty, to not give the final plugin an empty tags property.

dimitropoulos replied:
the next PR does a lot of work with tags, take a look there and see if your opinion's changed any. #3428

develohpanda commented on packages/openapi-2-kong/src/declarative-config/services.test.ts:
```diff
@@ -142,25 +149,28 @@ describe('services', () => {
-          parameter_schema: 'operation',
-        },
-      };
+      if (spec.paths['/cats'].post) {
```
Hmmm I generally avoid conditionals in unit tests because they can start causing false positives, but I assume this is needed for types to resolve correctly?

Is there another approach we can use here, or add `else` cases to fail the test if the conditional is not met?

dimitropoulos replied:
![Screenshot_20210526_103126](https://user-images.githubusercontent.com/15232461/119678475-89c25a00-be0d-11eb-951a-881cf255afc5.png)

Yep, exactly.  We can cast.  I'll do so in the next PR.

dimitropoulos replied:
989adc7cf0a71a71ebf7b70187ce21d793b6557a

develohpanda commented on packages/openapi-2-kong/src/declarative-config/services.ts:
```diff
@@ -159,6 +159,6 @@ export function generateRouteName(
   // replace all `/` with `-` except the ones at the beginning or end of a string
   const replacedRoute = routePath.replace(/(?!^)\/(?!$)/g, '-');
-  const pathSlug = generateSlug(pathItem['x-kong-name'] || replacedRoute);
+  const pathSlug = generateSlug(pathItem[xKongName] || replacedRoute);
```
I love this approach 💯

dimitropoulos replied:
:)

develohpanda commented on packages/openapi-2-kong/src/declarative-config/utils.ts:
```diff
@@ -1,7 +1,7 @@
 
 /** This function is written in such a way as to allow mutations in tests but without affecting other tests. */
-export const getSpec = (): OpenApi3Spec =>
+export const getSpec = (overrides: Partial<OpenApi3Spec> = {}): OpenApi3Spec =>
```
👍🏽

develohpanda commented on packages/openapi-2-kong/src/types/openapi3.ts:
```diff
@@ -1,39 +1,40 @@
-    }
-  }
+      verbose_response?: boolean;
+      parameter_schema?: 'global' | 'path' | 'operation';
```
This property doesn't seem right at all, but it also has a `bodySchema` which is defined under the `DcPluginConfig`.

Documentation here: https://docs.konghq.com/hub/kong-inc/request-validator/#parameters

dimitropoulos replied:
while I agree with you and the docs agree with you, the code (ours) does not: https://github.com/Kong/insomnia/blob/develop/packages/openapi-2-kong/src/declarative-config/services.test.ts#L140-L262

Let's follow up on this separately, like you're saying.  The current type simply satisfies the existing code.  If the existing code is wrong, I'm a full supporter of fixing :)

develohpanda replied:
Ya current code would be wrong (atleast in unit tests). 👍
